### DISCUSSION TITLE
Modularize RTK App

### DISCUSSION THREAD

**Discussion by sam-mfb** [reduxjs/redux-toolkit#1845] [Q&A] [answered]
This is re-post (at Mark's suggestion) of [this SO question](https://stackoverflow.com/questions/70394461/modularize-a-redux-toolkit-application/70405059#70405059), with some updates

### The Question
Is it possible to separate out the feature of an RTK-based application that depend on different slices of a the redux store into separate node packages? Assuming so, what is the best way to do that?

### Background
We have a large, and growing, app that is based around Redux Toolkit. Where possible we try to separate parts of the application into their own node packages. We find there are a lot of benefits to doing this, including:

- Maintainability of codebase
- Fine-grained control over intra-application dependencies
- Testability

It's easy enough to do this for cross-cutting things, like logging, http requests, routing, etc. But we would like to go further and modularize the "features" of our app. For example, have the "address book" feature of our application live in a different module than, say, the "messages" feature, with them all composed together via an "app" package.

The benefits we see here are ones we have found in other codebases and have been discussed in other places. (E.g., [here](https://www.pointfree.co/episodes/ep171-modularization-part-1) for iOS). But, in brief: (1) you can see and control intra-app dependencies. For example, you can easily see if the "messages" feature depends on the "address book" feature and make explicit decisions about how you will expose the one feature to the other via what you export; (2) you can build fully testable sub-parts of the app by simply having a "preview" package that only composes in the things you want to test, e.g., you could have a "contact app" package that only depends on the "contact" feature for building and testing just that; (3) you can speed up CI/CD times by not needing to compile (TS/babel), pack/minify, and unit test every part; (4) you can utilize various analytics tools to get more fine-grained pictures of how each feature is developing.

There may well be other ways to achieve these things, and some may disagree with the premise that this is a good way to do it. That's not the focus of the question, but I'm open to the possibility it may be the best answer (e.g., some one with significant Redux experience may explain why this is a bad idea).

### The Problem
We've struggled to come up with a good way to do this with Redux Toolkit. The problem seems to boil down to -- is there a good way to modularize (via separate node packages) the various "slices" used in RTK? (This may apply to other Redux implementations but we are heavily invested in RTK).

It's easy enough to have a package that exports the various items that will be used by the redux store, i.e., the slice state, action creators, async thunks, and selectors. And RTK will then compose those very nicely in the higher-level app. In other words, you can easily have an "app" package that holds the store, and then a "contacts" package that exports the "contacts" slice, with its attendant actions, thunks, selectors, etc.

The problem comes if you also want the components and hooks that use that portion of slice to live in the same package as the slice, e.g., in the "contacts" package. Those components/hooks will need access to the global dispatch and the global useSelector hook to really work, but that only exists in the "app" component, i.e., the feature that composes together the various feature packages.

### Possibilities Considered
We could export the global dispatch and useSelector from the "higher" level "app" package, but then our sub-components now depend on the higher level packages. That means we can no longer build alternate higher level packages that compose different arrangements of sub packages.

We could use separate stores. This has been discussed in the past regarding Redux and has been [discouraged](https://stackoverflow.com/questions/33619775/redux-multiple-stores-why-not), although there is [some suggestion](https://redux.js.org/faq/store-setup) it might be OK if you are trying to achieve modularization. These discussions are also somewhat old.

### Current thinking

In-line with Mark's suggestion on the SO question, we are thinking the best way to approach it may be to export an `initialize` function from each of the subprojects that the the app project calls, that basically injects the global `dispatch` and the `useSelector` functions from the global store.  This would allow a kind of dependency injection, so there does not need to be a tight coupling of the parent app to sub projects.  There may be some trickiness around typing but conceptually I feel like this should work.  We are going to work on a proof of concept and see how it goes.

Any thoughts or suggestions welcome.

**Comment by sam-mfb** [comment]
For completeness, here's Mark's post on SO:

This question has come up in other contexts, most notably how to write selector functions that need to know where a given slice's state is attached to the root state object. Randy Coulman had an excellent and insightful series of blog posts on that topic back in 2016 and a follow-up post in 2018 that cover several related aspects - see [Solving Circular Dependencies in Modular Redux ](https://randycoulman.com/blog/2018/06/12/solving-circular-dependencies-in-modular-redux/)for that post and links to the prior ones.

My general thought here is that you'd need to have these modules provide some method that allows injecting the root dispatch or asking the module for its provided pieces, and then wires those together at the app level. I haven't had to deal with any of this myself, but I agree it's probably one of the weaker aspects of using Redux due to the architectural aspects.

For some related prior art, you might want to look at these libraries:

https://github.com/ioof-holdings/redux-dynostore (deprecated / unmaintained, but relevant)
https://github.com/microsoft/redux-dynamic-modules (also may be unmaintained at this point - still seems to rely on React-Redux v5)
https://github.com/fostyfost/redux-eggs (brand new - the author just posted this on the RTK "Discussions" section recently)

Might also be worth filing this same question over in the RTK "Discussions" area as well so we can talk about it further.

**Reply by markerikson** [reply]
For reference I also linked the SO question on Twitter and got a couple responses, although they were more along the lines of "we code-split some reducers" rather than "publishing an entire feature as a package by itself":

https://twitter.com/acemarke/status/1472240035458170888

**Reply by sam-mfb** [reply]
Right.  I should clarify for others looking at this thread that this is different (though perhaps related) to code-splitting or runtime injection.  We are looking to be able to develop/test/build separately.  Currently we are Ok with the app itself running as a monolith (or at least, we treat that as a separate question).

**Comment by sam-mfb** [comment]
OK, here is a proof of concept for the dependency injection model.  https://github.com/sam-mfb/rtk-split

It works, but there are a couple of places where I'm hiding typescript issues.  I think these issues are not fundamental, however, but just related to an issue typescript has with pnpm symlinks.  More info on that at the end.

This basically just takes the default redux-typescript template from create-react-app and breaks out the "counter" feature as a separate package.  The key thing that lets it work is the following bits in `counterSlice.ts`:

```typescript
const localStore = configureStore({
  reducer: { counter: counterSlice.reducer },
});

type RootState = { counter: CounterState } & Record<string, any>;
type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
type SliceDispatch = typeof localStore.dispatch;

export let useSliceSelector: TypedUseSelectorHook<
  { counter: CounterState } & Record<string, any>
> = useSelector;

export let useSliceDispatch = () =>
  useDispatch<SliceDispatch & ThunkDispatch<any, any, any>>();

export const initializeSlicePackage = (
  useAppDispatch: typeof useSliceDispatch,
  useAppSelector: typeof useSliceSelector
) => {
  useSliceDispatch = useAppDispatch;
  useSliceSelector = useAppSelector;
};
```
Basically I'm typing RootState (and the derivative types) as any composed state that has *at least* the state of the slice i'm using in the package (and same for the dispatch).  Those are then used throughout this package (i.e. in `Counter.tsx`), but... there's also the `initializeSlicePackage()` that's exported that allows these "local" objects to be overridden by the *global* `useAppDispatch` and `useAppSelector` from the parent `app`.  That's done by a simple call in `hooks.ts` in the parent app.

This all works and is basically what I was hoping to achieve.  I think the typing could be improved.  Currently to get the type for the local dispatch I'm actually creating it in Javascript.  I feel like there should be a way to infer it entirely in Typescript without any runtime, but I haven't spent enough time there.

(The TS problem in the current version is masked by typing `counterSlice` as `any`.  If you remove that assertion, TS complains but I don't think this is a fundamental problem with the inference.  It's complaining about being unable to name the inferred type because of a symlinked package, which I think is a pnpm/typescript interaction issue.  See, e.g., https://github.com/microsoft/TypeScript/issues/42873   That's also why I have a `@ts-ignore` in `hooks.js`. I obviously need to fix this issue for this to be a viable solution but I don't think its a problem with the overall approach.)

**Comment by sam-mfb** [comment]
I fixed the TS problem (it just required an explicit import of immer so it could use WriteableDraft in the declaration).  

That's updated in the repo now.

**Comment by sam-mfb** [comment, marked as answer]
I cleaned up the typing and added some documentation.  Here's the full solution, which is also on [the repo](https://github.com/sam-mfb/rtk-split) (suggestions and improvements welcome):
```typescript
// RootStateInterface is defined as including at least this slice and any other slices that
// might be added by a calling package
type RootStateInterface = { counter: CounterState } & Record<string, any>;

// A version of AppThunk that uses the RootStateInterface just defined
type AppThunkInterface<ReturnType = void> = ThunkAction<
  ReturnType,
  RootStateInterface,
  unknown,
  Action<string>
>;

// A version of use selector that includes the RootStateInterface we just defined
export let useSliceSelector: TypedUseSelectorHook<RootStateInterface> =
  useSelector;

// This function would configure a "local" store if called, but currently it is
// not called, and is just used for type inference.
const configureLocalStore = () =>
  configureStore({
    reducer: { counter: counterSlice.reducer },
  });

// Infer the type of the dispatch that would be needed for a store that consisted of
// just this slice
type SliceDispatch = ReturnType<typeof configureLocalStore>["dispatch"];

// AppDispatchInterface is defined as including at least this slices "local" dispatch and
// the dispatch of any slices that might be added by the calling package.
type AppDispatchInterface = SliceDispatch & ThunkDispatch<any, any, any>;

export let useSliceDispatch = () => useDispatch<AppDispatchInterface>();

// Allows initializing of this package by a calling package with the "global"
// dispatch and selector hooks of that package, provided they satisfy this packages
// state and dispatch interfaces--which they will if the imported this package and
// used it to compose their store.
export const initializeSlicePackage = (
  useAppDispatch: typeof useSliceDispatch,
  useAppSelector: typeof useSliceSelector
) => {
  useSliceDispatch = useAppDispatch;
  useSliceSelector = useAppSelector;
};
```

**Reply by markerikson** [reply]
Nice! I particularly like the "make a function that _would_ return a store to get its type, but don't actually call it" trick :)

**Reply by sam-mfb** [reply]
One thing I'm still thinking about on this solution is how to handle accessing state or dispatching actions across features.  For example, what if the "counter" package needs some state from a a different feature, say, a "todo" feature.  Or wants to change the state of the "todo" feature, i.e., dispatch an action.  It will obviously then depend on "todo" and import it, but what's the implementation?

I see two options: 

A.  My inclination is to say the "todo" feature should expose its own functions that let other packages do that, e.g., a `getToDo(item: number): string` function or a `setToDo(item:number, toDo: string):void` function.  And then, internally in the 'todo' package have those functions wrap the relevant calls to `useSelector` or `dispatch`.  That way the fact that "todo" is also using redux is basically hidden from the "counter" function, i.e., it's an implementation detail.  That seems cleanest, most flexible, and most consistent with the loose coupling I'm trying to achieve.

B.  Another alternative, though, would be that "counter" could directly import "todo"'s selectors and action creators, and then pass them on to its, initialized, `dispatch` or `useSelector`.  That would work (I think) but I don't love it, as it requires the "counter" package to know that the "todo" package also has a redux dependency.  You then couldn't change that down the road, if, say, you decided the "todo" package didn't need redux after all. You also wouldn't have any type safety or Intellisense help when passing your "todo" selectors or actions from within "counter", at least without further manipulation of the dispatch and state types inside "todo."

This also highlights a slight weakness in the current implementation, namely, that the `dispatch` and `useSelector` passed to "counter" from "app" are fully global.  Accordingly, even if you implement a solution like A above, nothing prevents a consumer in "counter" from passing in arbitrary actions or selectors (which the other feature components will likely expose since the "app" component needs them).  I could imagine one way to handle that would be to have the "app" pass in a partially disabled or limited `useAppDispatch` and `useAppSelector` to the initialize function that checks that actions/selectors are permitted before passing them on to the "real" hooks.  That would effectively constrain the "counter" package to stay within its domain, and ensure that devs followed A, if that was the chosen architecture.  I may play with adding something like that to the current solution.

**Reply by lemke-ethan** [reply]
Took a crack at [(A)](https://github.com/sam-mfb/rtk-split/pull/4).

**Reply by markerikson** [reply]
Just out of curiosity, does this new https://github.com/fostyfost/redux-eggs library address any of the problems that are being solved here?

**Reply by sam-mfb** [reply]
@lemke-ethan's solution is nice (full disclosure, ethan and I work together).  the nice thing he realized is that you don't actually need to widen the state and dispatch function types in "counter" because, since they are functions, they are in contravariant positions so it's no problem for the "main-app" to pass "wider" hooks.  and then as a side benefit you get narrower types within "counter" which--to some extent---prevents a consumer inside "counter" from going outside that feature.

I say "to some extent" because (a) it's only typescript, so you could just override the errors with assertions, and (b) the typing only restricts the selectors you pass, not the actions you pass to dispatch.  On (b) it looks to me like that's generally true of `dispatch`, even if you don't modularize, i.e., that it gets types as `Dispatch<AnyAction>` so there is no restriction on what form of action you pass.  That probably makes sense for a single package, since any "bad" actions are just no-ops and state doesn't change.  A little different here where the idea would be to block (at least in a Typescript sense) actions that are "good-but-from-the-wrong-feature".

Ethan and I talked about whether these checks should be implemented in runtime as well as TS.  Our inclination is no as you would effectively be adding a bit of overhead (i.e., an extra switch/loop to check if an action is valid before passing to the reducer) solely to enforce some code hygiene.

**Reply by sam-mfb** [reply]
I only briefly looked at redux-eggs.  I'd like to take a closer look as it sounds interesting, and code-splitting/injection is sort of a natural outgrowth of this modularization.

**Reply by markerikson** [reply]
Yeah, per the `dispatch` thing, we actually specifically recommend _not_ trying to limit what can be passed to `dispatch` at the type level - it's entirely legal to dispatch any action at any time.  See https://redux.js.org/usage/usage-with-typescript#avoid-action-type-unions .

**Reply by omdxp** [reply]
what if you have entity adapters within some slices?

**Reply by markerikson** [reply]
@omdxp I don't think I understand the question - can you clarify?

**Reply by sam-mfb** [reply]
if the question is whether entity adapters would work with this modularization approach, i haven't tried, but I don't see why it wouldn't.  As I understand it entity adapters are, in essence, just helpers that generate selectors and case reducers, all of which happens at the slice level.  It should therefore work fine with this approach which basically just involves injecting the useSelector and useDispatch hooks from a parent app in to the child package but is indifferent as to how you generate the selectors or actions that get passed to those hooks.

**Reply by omdxp** [reply]
@markerikson it's how @sam-mfb explained. I just want to see how this works when you have something like this:
```ts
const initialState = documentsAdapter.getInitialState<LearnScreenState>({
  status: "idle",
  error: null,
});
```
where the state of the slice will contains the following properties: `status`, `error`, `entities` and `ids`.

In this example, to select the documents we can use:
```ts
export const { selectAll: selectDocuments } = documentsAdapter.getSelectors(
  (state: RootState) => state.learnScreen
);
```

**Reply by sam-mfb** [reply]
So if you are looking at the example above, what you have should work fine, except you would type your selector's state with `RootStateInterface` rather than `RootState`, i.e., tell it that the root state will _at least_ include your slice state, rather than giving it the full RootState that will be composed at the parent app (and that your child package doesn't need to know about).

**Reply by omdxp** [reply]
@sam-mfb I would like to include `documents` state within the `useSliceSelector` which can be selected with `selectDocuments` selector. if I use `useSliceSelector` as mentioned above I wouldn't get all the `documents` but instead I'd get the `entities` and `ids` which is not ergonomic to map through the `ids` and return the `document` for each one

**Reply by sam-mfb** [reply]
Would you be able to create a fork off of [this repo](https://github.com/sam-mfb/rtk-split) that demonstrates the problem you are having?  I have a fair bit of experience with doing this sort of modularization with RTK, but not much experience with EntityAdapter, so I'm just having a little trouble following the problem in the abstract.

**Reply by omdxp** [reply]
Ok, will do it later

**Comment by FaberVitale** [comment]
Are you aware of the react-boilerplate approach used in [react-bolierplate-cra-template](https://github.com/react-boilerplate/react-boilerplate-cra-template)?

It injects dynamically reducers and sagas.
It is based on this package: https://www.npmjs.com/package/redux-injectors

It probably deserves a closer look to see if there good ideas we could take inspiration from.

**Reply by markerikson** [reply]
I think the big difference there is that `react-boilerplate` is "just" separating things into different folders in one app, while the OP wants to actually maintain each feature as an entirely separate package.

**Comment by asherccohen** [comment]
Looking into his topic too.

We're modularizing our apps in a large monorepo and need to extract features in separate libraries.

We've currently tried all the packages and all have their trade offs.

In the near future we'll also explore module-federation, so being able to import/inject slices and APIs even at runtime is fundamental.

I'll keep updating this thread with our findings.

**Comment by sam-mfb** [comment]
BTW, in the year since I made this post, we have fully modularized our application's features using this strategy.  Each feature it its own npm package and has its own redux slice that is entirely internal to it, and all the features get composed in a parent app using the dependency injection method explained here.  This does not, however, happen at runtime--it's all part of the build process.

**Reply by umashankar-pardhi-amla** [reply]
@sam-mfb : Can you please create the Repo and share here this will help us lot

**Reply by sam-mfb** [reply]
The repo is the same one we posted above.  [Here it is again.](https://github.com/sam-mfb/rtk-split) My only point in this post was that we have extended this strategy to our working production code (which is proprietary).  But the basic approach and strategy is the same.  

If there's interest, I can go through it and see if there are any minor tweaks to add based on our production code, but I'm a bit underwater right now and won't have time for a while.  In the meantime though, if you have specific questions about trying out this approach feel free to post and I will try to answer.  We have really liked having this modularity.

**Reply by upardhi** [reply]
What I am trying to achieve is I can register my slice/reducer to their only package, and also we can access the counter in multiplier directly through the store.

I have multiple A, B, C, D ..... Z features and some customers are using A feature only, Some Are using B and C only so I want to make it modular so that I can not mix the dependencies at the time of application load

I do not want to register all slices initially .. Thoughts?


![image](https://user-images.githubusercontent.com/6356988/236111841-bf18cde1-dca2-471f-886b-58a05c3025b6.png)

**Reply by sam-mfb** [reply]
This functionality is actually there in the example app, but it is hidden (intentionally) by the typing.

The useAppSelector and useAppDispatch hooks injected into Multiplier actually can access the full store as a matter of javascript.  That is, they are just the useSelector and useDispatch hooks generated by react-redux--we haven't altered them in any way. The only reason they appear not to be able to is because of the narrow Typescript typings we put on them inside the multiplier package to prevent devs from accidentally going outside of feature scope.  I.e., we explicitly do not want our dev accessing one feature from inside another. 

Since you want to be able to go out of feature scope and access the counter store directly, you need to broaden those typings.  

Making the broader types is a little tricky as you can't just use these lines like we do in the Main App:
```
export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
```
Because your Multiplier app won't "know" about the shape of the Counter state and there isn't any way to "inject" types down from the Main App package into the Multiplier packages. (Since injection is a runtime concept, and typescript is compile time.).  Off the top of my head, the way I would handle that would be to make type-only package, let's call it "counter-types" that will be imported as dev-dependency by "Multiplier" and will be used to get the shape of the counter store and selectors for purposes of constructing the store and selector types inside multiplier.  The idea being you want to give Multiplier access to Counter store's types without giving it access to its code.  You'll then have to compose the AppDispatch and RootState types manually inside your Multiplier package but it shouldn't be overly difficult. 

All that said, I would humbly suggest, based on our experience, that you will find the structure of our repo much easier to maintain long term than the one you are proposing.  The "Multiplier" package you are proposing will be end up leading to the same kind of maintenance headaches that this modularity approach we are proposing is meant to avoid.  I.e., every time you change some selector or reducer interface in Counter you have to go into Multiplier and change code there as well.  Multiply that by x number of features and you are back in feature dependency hell :).  If you want to be able to provide some users with feature A and some users with features A B, I would just construct two separate Main Apps.  The beauty of this solution is that the Main Apps end up being super simple to write since all they are doing is composing together multiple features.  In fact, in our repo, we always create little preview Main Apps that only use one feature for easy development and testing.  That's just my 2 cents.
